pkg/validation/admission_test.go

```diff
@@ -0,0 +1,128 @@
+
+	metal3apis "github.com/metal3-io/baremetal-operator/pkg/apis"
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/pkg/apis/metal3/v1alpha1"
+)
+
+const (
+	namespace                 string = "test-namespace"
+	defaultBMCCredentialsName string = "bmc-creds-valid"
+	defaultBootMACAddress     string = "00:00:00:00:00"
+	defaultBMCAddress         string = "ipmi://192.168.122.1:6233"
+)
+
+func init() {
+	logf.SetLogger(logf.ZapLogger(true))
+	// Register our package types with the global scheme
+	metal3apis.AddToScheme(scheme.Scheme)
+}
+
+func newHost(bmcAddress, bmcCredentialsName, bootMACAddress string) metal3v1alpha1.BareMetalHost {
+	return metal3v1alpha1.BareMetalHost{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BareMetalHost",
+			APIVersion: "metal3.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "myhost",
+			Namespace: namespace,
+		},
+		Spec: metal3v1alpha1.BareMetalHostSpec{
+			BMC: metal3v1alpha1.BMCDetails{
+				Address:         bmcAddress,
+				CredentialsName: bmcCredentialsName,
+			},
+			BootMACAddress: bootMACAddress,
+		},
+	}
+}
+
+func newDefaultHost() metal3v1alpha1.BareMetalHost {
+	return newHost(defaultBMCAddress, defaultBMCCredentialsName, defaultBootMACAddress)
+}
+
+func TestCanBeAdmitted(t *testing.T) {
+	testCases := []struct {
+		Scenario string
+		Host     metal3v1alpha1.BareMetalHost
+		Expected error
+	}{
+		{
+			Scenario: "BMC address is not unique",
+			Host:     newHost(defaultBMCAddress, "", ""),
+			Expected: apierrors.NewInvalid(
+				metal3v1alpha1.SchemeGroupVersion.WithKind("BareMetalHost").GroupKind(),
+				"myhost",
+				field.ErrorList{
+					field.Invalid(
+						field.NewPath("spec", "bmc", "address"),
+						defaultBMCAddress,
+						"is not unique",
+					),
+				}),
+		},
+		{
+			Scenario: "BMC address is unique",
+			Host:     newHost("unique", "", ""),
+			Expected: nil,
+		},
+		{
+			Scenario: "BootMACAddress is not unique",
+			Host:     newHost("", "", defaultBootMACAddress),
+			Expected: apierrors.NewInvalid(
+				metal3v1alpha1.SchemeGroupVersion.WithKind("BareMetalHost").GroupKind(),
+				"myhost",
+				field.ErrorList{
+					field.Invalid(
+						field.NewPath("spec", "bootMACAddress"),
+						defaultBootMACAddress,
+						"is not unique",
+					),
+				}),
+		},
+		{
+			Scenario: "BootMACAddress is unique",
+			Host:     newHost("", "", "unique"),
+			Expected: nil,
+		},
+		{
+			Scenario: "BMC credentials name is not unique",
+			Host:     newHost("", defaultBMCCredentialsName, ""),
+			Expected: apierrors.NewInvalid(
+				metal3v1alpha1.SchemeGroupVersion.WithKind("BareMetalHost").GroupKind(),
+				"myhost",
+				field.ErrorList{
+					field.Invalid(
+						field.NewPath("spec", "bmc", "credentialsName"),
+						defaultBMCCredentialsName,
+						"is not unique",
+					),
+				}),
+		},
+		{
+			Scenario: "BMC credentials name is unique",
+			Host:     newHost("", "unique", ""),
+			Expected: nil,
+		},
```

Comment:
It looks like we could use a test for multiple hosts with none of the optional fields set.